### PR TITLE
macOS and Windows support

### PR DESCRIPTION
**Pull request by hpwxf**
(cf issue #28)

Refactor travis config into *sub* files.
MacOS support is very close to the initial Linux support.
Windows support is ready for production but disabled since there are a few bugs to fix before (windows settings are as comments in `.travis-ci.yml`).
Windows compilation failure example: https://travis-ci.com/github/hpwxf/carma/builds/174076918

I will add comments (below) about these changes.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -36,7 +36,11 @@ OPTION(BUILD_EXAMPLES "Build examples for carma" OFF)
 # ##############################################################################
 #                                   TARGET                                     #
 # ##############################################################################
-SET(PROJECT_RELEASE_FLAGS "-march=native" "-mtune=native")
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" 
+        OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang" 
+        OR CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
+    SET(PROJECT_RELEASE_FLAGS "-march=native" "-mtune=native")
```
Comment (hpwxf) [inline CMakeLists.txt:42]: I only use it for GCC, Clang, and AppleClang.
Moreover, these options are good to get the best code optimization but it reduces the portability of the generated binaries.

File: .travis.yml [hunk 2]
```diff
@@ -71,86 +71,53 @@ matrix:
...
-      -Dsonar.host.url=https://sonarcloud.io \
-      -Dsonar.cfamily.build-wrapper-output=bw-output \
-      -Dsonar.login=890640f8296b91c98ccfa8ec2545220d68a4edb6
-  else
-    make
-    cd ..
-  fi
-  set +ex
+  - |
+    #!/usr/bin/env bash
```
Comment (hpwxf) [inline .travis.yml:100]: she-bang is required for macOS since default shell is zsh (to prevent possible incompatibilities)

**Comment by hpwxf** [comment]
* I prefer to use one variable to express the compiler type and its version. I did it with `COMPILER` variable (ex: `COMPILER=gcc-7`)
* `make` commands have been replaced by `cmake` equivalent command (more portable with Windows)
* `before_install.sh` script is about things that should be so common on many OS (Python and curl). This requirement is fundamental to have a common installation process on the 3 targeted OSes.

Comments in code:
* https://github.com/RUrlus/carma/pull/32/files#r449050314
* https://github.com/RUrlus/carma/pull/32/files#r449048536

**Comment by hpwxf** [comment]
Don't merge now. I'm working to fix the Windows compilation error. We can include this in this PR.

After that, you can merge this PR using the *squash* option of GitHub.

File: CMakeLists.txt [hunk 3]
```diff
@@ -36,7 +36,15 @@ OPTION(BUILD_EXAMPLES "Build examples for carma" OFF)
...
+elseif (CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+    # workaround like https://github.com/nlohmann/json/issues/1408
+    # to avoid error like: carma\third_party\armadillo-code\include\armadillo_bits/arma_str.hpp(194): error C2039: '_snprintf': is not a member of 'std' (compiling source file carma\tests\src\bindings.cpp) 
+    ADD_DEFINITIONS(-DHAVE_SNPRINTF)
```
Comment (hpwxf) [inline CMakeLists.txt:46]: Requires to prevent from bad #define in armadillo

**Comment by hpwxf** [comment]
Bugs in Windows are not easy to fix: cf https://travis-ci.com/github/hpwxf/carma/jobs/357100178
In this PR Windows CICD is ready but not enabled.

**Comment by RUrlus** [comment]
Happy to merge like this, thanks for all the work.

**Pull request merged**